homeassistant/components/tellduslive/__init__.py

```diff
@@ -70,8 +70,8 @@
 
     client = TelldusLiveClient(hass, entry, session)
     hass.data[DOMAIN] = client
-
-    await client.update()
+    hass.async_create_task(async_add_hubs(hass, client, entry.entry_id))
+    hass.async_create_task(client.update())
```

Comment:
Inside the `client.update` method there's a call to self._client.update`. That is doing I/O. That needs to be scheduled on the executor thread pool with `hass.async_add_executor_job`.

Reply:
Maybe we should await this to make sure no other threads are spawned for client updates before the first update is done?

Reply:
That is not the case, the problem is that `gather` "forks" too many processes [here](https://github.com/home-assistant/home-assistant/pull/19180/files#diff-6182200c1051279a736e0dd1a80942d9R202).

Reply:
Ok, so a simple fix could be to remove `asyncio.gather` and just await each discover in a loop. Would that work?

Reply:
Yes, but can't we throttle gather so we send say 5 tasks concurrently?

Reply:
We could add a semaphore to the client that the discover coroutine acquires.

https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore

Reply:
I've made some tests:

* Current implementation (yields errors and missing entities) **gather: 10.14268684387207 s**
* Current implementation with `pool_block` **gather: 1.1395719051361084 s**
* await for each discover **1.0583820343017578 s**
```python
for d_id in new_devices:
    await self._discover(d_id)
```
* await for discover split in chunks of 10 **1.065885066986084 s**
```python
for i in range(0, len(new_devices), 10):
    await asyncio.gather(*[self._discover(d_id) for d_id in list(new_devices)[i:i + 10]])
```